sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -90,7 +90,7 @@ case class AdaptiveSparkPlanExec(
   // Exchange nodes) after running these rules.
   private def queryStagePreparationRules: Seq[Rule[SparkPlan]] = Seq(
     ensureRequirements
-  )
+  ) ++ context.session.sessionState.queryStagePrepRules
```

Comment:
Should we do this after ensure requirements? The `queryStagePrepRules` might break requirements.

Reply:
The custom rules may need to see exchange nodes as well. We can do validation at the end to make sure exchange requirements are still required.